server/org.eclipse.emfcloud.ecore.glsp/src/main/java/org/eclipse/emfcloud/ecore/glsp/operationhandler/EcoreDeleteOperationHandler.java

```diff
@@ -21,18 +30,33 @@
 public class EcoreDeleteOperationHandler extends DeleteOperationHandler {
 	@Override
 	protected boolean delete(String elementId, GModelIndex index, GraphicalModelState graphicalModelState) {
-		
 		super.delete(elementId, index, graphicalModelState);
 		EcoreModelState modelState = EcoreModelState.getModelState(graphicalModelState);
 		
-		modelState.getIndex().getSemantic(elementId).ifPresent(element -> {
+		Optional<EObject> semantic = modelState.getIndex().getSemantic(elementId);
+		Optional<NotationElement> notation = modelState.getIndex().getNotation(elementId);
+		
+		semantic.ifPresent(element -> {
 			if(element instanceof EReference && ((EReference) element).getEOpposite() != null) {
 				EcoreUtil.delete(((EReference) element).getEOpposite());
 				modelState.getIndex().getBidirectionalReferences().remove(EcoreEdgeUtil.getStringId((EReference)element));
 			}
+			if (element instanceof EClassifier) {
```

Comment:
Maybe I missed something, but wouldn't it suffice to just call
`EcoreUtil.delete(element, true)`
Imho that would delete the element and all incoming references and with `recursive=true` it would do the same for all children?

Reply:
Yes, I think you missed the PR Comment :) :

> The second commit improves the delete action to avoid leaving dangling hrefs. Note that it wasn't sufficient to just call EcoreUtil::delete in this case, because untyped EReferences are currently not well handled and caused unrecoverable errors on the client side; so I currently delete all EAttributes/EReferences typed with the deleted classifiers (It really makes sense for EReferences, since the edge is deleted as well; for EAttributes it's a bit weirder, as seemingly unrelated EAttributes might disappear when deleting an EDataType or EEnum)